reloadly/giftcards_test.go

```diff
@@ -0,0 +1,355 @@
+
+	assert.NotNil(t, err)
+	assert.Equal(t, 400, err.(APIError).StatusCode)
+}
+
+func TestDiscounts(t *testing.T) {
+	dat, _ := ioutil.ReadFile("test/discounts.json")
+	discounts := string(dat)
+
+	ts, _ := TestServer(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, "/discounts", r.URL.Path)
+
+		w.Header().Set("Content-Type", "application/com.reloadly.giftcards-v1+json")
+		fmt.Fprintf(w, discounts)
+	})
+
+	svc := &Service{BaseUrl: ts.URL, Client: &http.Client{}}
+	res, err := svc.GiftCards().Discounts(1, 100)
+
+	assert.Nil(t, err)
+	assert.Equal(t, "1-800-PetSupplies", res.Content[0].Product.ProductName)
+	assert.Equal(t, int64(2), res.Content[1].Product.ProductID)
+}
+
+func TestDiscountsPagination(t *testing.T) {
+	ts, _ := TestServer(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, "/discounts", r.URL.Path)
+		assert.Equal(t, "page=2&size=1", r.URL.RawQuery)
+	})
+
+	svc := &Service{BaseUrl: ts.URL, Client: &http.Client{}}
+	_, err := svc.GiftCards().Discounts(2, 1)
+
+	assert.NotNil(t, err)
+}
+
+func TestDiscountByProduct(t *testing.T) {
+	ts, _ := TestServer(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, "/products/2/discounts", r.URL.Path)
+
+		w.Header().Set("Content-Type", "application/com.reloadly.giftcards-v1+json")
+		fmt.Fprintf(w, `{"product":{"productId":2,"productName":"Amazon UK","countryCode":"GB","global":false},"discountPercentage":0.7}`)
+	})
+
+	svc := &Service{BaseUrl: ts.URL, Client: &http.Client{}}
+	res, err := svc.GiftCards().DiscountByProduct(2)
+
+	assert.Nil(t, err)
+	assert.Equal(t, "Amazon UK", res.Product.ProductName)
+	assert.Equal(t, 0.7, res.DiscountPercentage)
+}
+
+func TestDiscountByProductReturnsNotFoundError(t *testing.T) {
+	ts, _ := TestServer(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, "/products/25/discounts", r.URL.Path)
+
+		w.WriteHeader(404)
+		fmt.Fprintf(w, `{"timeStamp":"2021-11-16 17:28:15","message":"Commission resource not found","path":"/products/2000/discounts","errorCode":null,"infoLink":null,"details":[]}`)
+	})
+
+	svc := &Service{BaseUrl: ts.URL, Client: &http.Client{}}
+	_, err := svc.GiftCards().DiscountByProduct(25)
+
+	assert.NotNil(t, err)
+	assert.Equal(t, 404, err.(APIError).StatusCode)
+}
+
+func TestTransactions(t *testing.T) {
+	dat, _ := ioutil.ReadFile("test/transactions.json")
+	transactions := string(dat)
+
+	ts, _ := TestServer(func(w http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, "/reports/transactions", r.URL.Path)
+
+		w.Header().Set("Content-Type", "application/com.reloadly.giftcards-v1+json")
+		fmt.Fprintf(w, transactions)
+	})
+
+	svc := &Service{BaseUrl: ts.URL, Client: &http.Client{}}
+	res, err := svc.GiftCards().Transactions(1, 100)
+
+	assert.Nil(t, err)
+	assert.Equal(t, "INR", res.Content[0].CurrencyCode)
+	assert.Equal(t, "yeahyeah", res.Content[0].CustomIdentifier)
+}
+
+func TestTransactionsPagination(t *testing.T) {
```

Comment:
Wow, supporting transactions as well! You're really trying to compete with the official go sdk, eh?